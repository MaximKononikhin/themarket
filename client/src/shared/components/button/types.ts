import { ButtonHTMLAttributes, ReactNode } from "react";

export interface ButtonProps extends ButtonHTMLAttributes<HTMLButtonElement> {
    /** Контент кнопки **/
    children?: ReactNode;
    /** id **/
    testId?: string;
    className?: string;
    wide?: boolean;
}
