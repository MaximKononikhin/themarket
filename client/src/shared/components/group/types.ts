export type GroupProps = {
    gap?: number;
    direction: "column" | "row";
    alignItems?: "center" | "flex-start" | "flex-end" | "end";
    width?: string;
    className?: string;
    margin?: string;
    testId?: string;
};
