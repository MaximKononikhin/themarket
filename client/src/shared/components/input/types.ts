import { InputHTMLAttributes, ReactNode } from "react";

import { Type } from "../typography/type";

export interface InputProps
    extends InputHTMLAttributes<HTMLInputElement | HTMLTextAreaElement> {
    /** Тип инпута **/
    design?: "input" | "textarea";
    /** Ширина инпута **/
    width?: "wide" | "normal" | "small";
    /** Лейбл **/
    label?: string | ReactNode;
    /** Плейсхолдер **/
    placeholder?: string;
    /** Текст ошибки **/
    error?: string;
    /** Тип ошибки **/
    errorType?: "absolute" | "static";
    /** labelStyle **/
    labelStyle?: Extract<Type, "header-1" | "header-2">;
    /** стилизация **/
    className?: string;
    /** тест-айди **/
    testId?: string;
}
