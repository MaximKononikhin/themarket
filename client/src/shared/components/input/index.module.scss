.label {
    display: flex;
    flex-direction: column;
    gap: 10px;
    width: fit-content;
    width: 100%;
}

.label__wide {
    width: 100%;
}

.input__common {
    width: 386px;
    border: 1px solid #e1e3e0;
    border-radius: 8px;
    font-weight: 400;
    font-size: 15px;
    line-height: 19px;
    letter-spacing: -0.02em;
    color: #383838;
    padding: 12px 12px 13px 12px;
    outline: unset;
    text-overflow: ellipsis;
    white-space: nowrap;
    transition: all 0.25s ease;

    &::placeholder {
        font-weight: 400;
        font-size: 15px;
        line-height: 19px;
        letter-spacing: -0.02em;

        color: rgba(30, 29, 29, 0.3);
    }

    &:focus {
        border: 1px solid #cccccc;
    }
}

.socialInputContainer {
    position: relative;
}

.icon {
    position: absolute;
    left: 8px;
    bottom: 8px;
    z-index: 1;
    width: 28px;
    height: 28px;
}

.input__common_invalid {
    border: 1px solid #ee4a3c;

    &:focus {
        border: 1px solid #ee4a3c;
    }
}

.input__common_wide {
    width: 100%;
}

.input__common_small {
    width: 255px;
}

.input {
    @extend .input__common;
    max-height: 44px;
}

.textarea {
    @extend .input__common;
    min-height: 71px;
    resize: none;
    text-overflow: unset;
    white-space: pre-wrap;
}

.wrapper {
    position: relative;
}

.wrapper__wide {
    width: 100%;
}

.baseError {
    font-weight: 400;
    font-size: 13px;
    line-height: 17px;
    letter-spacing: -0.02em;
    color: #ee4a3c;
}

.error__absolute {
    @extend .baseError;
    position: absolute;
    bottom: -21px;
    left: 0;
}

.error__static {
    @extend .baseError;
    margin: 5px 0 0;
}
