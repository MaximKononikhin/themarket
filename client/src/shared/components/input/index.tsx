import React, { forwardRef } from "react";

import cn from "classnames";

import { Typography } from "@shared/components";
import { useAutosizeTextArea } from "@shared/lib/use-autosize-textarea";
import { useForwardRef } from "@shared/lib/use-forward-ref";

import styles from "./index.module.scss";
import { InputProps } from "./types";

export const Input = forwardRef<
    HTMLInputElement | HTMLTextAreaElement,
    InputProps
>((props, forwardedRef) => {
    const {
        value,
        onChange,
        width = "normal",
        label,
        placeholder,
        error,
        errorType = "absolute",
        labelStyle = "text-2",
        className,
        testId,
        design = "input",
        ...rest
    } = props;

    const ref = useForwardRef<HTMLInputElement | HTMLTextAreaElement>(
        forwardedRef
    );

    const elementProps = { value, onChange, placeholder, ...rest };

    if (design === "textarea" && typeof value === "string") {
        useAutosizeTextArea(ref.current as HTMLTextAreaElement, value);
    }

    const labelContent =
        typeof label === "string" ? (
            <Typography type={labelStyle}>{label}</Typography>
        ) : (
            label
        );

    const inputStyles = cn(className, {
        [styles.input]: true,
        [styles.input__common_invalid]: error,
        [styles.input__common_wide]: width === "wide",
        [styles.input__common_small]: width === "small",
    });

    const textAreaStyles = cn({
        [styles.textarea]: true,
        [styles.input__common_invalid]: error,
        [styles.input__common_wide]: width === "wide",
        [styles.input__common_small]: width === "small",
    });

    const labelStyles = cn({
        [styles.label]: true,
        [styles.label__wide]: width === "wide",
    });

    const wrapperStyles = cn({
        [styles.wrapper]: true,
        [styles.wrapper__wide]: width === "wide",
    });

    const errorStyles = cn({
        [styles.error__absolute]: errorType === "absolute",
        [styles.error__static]: errorType === "static",
    });

    const input = (
        <input
            ref={ref as React.LegacyRef<HTMLInputElement>}
            data-testid={testId}
            className={inputStyles}
            {...elementProps}
        />
    );

    const content =
        design === "input" ? (
            <>{input}</>
        ) : (
            <textarea
                ref={ref as React.LegacyRef<HTMLTextAreaElement>}
                className={textAreaStyles}
                {...elementProps}
            />
        );

    return (
        <div className={wrapperStyles}>
            <label className={labelStyles}>
                {labelContent}
                {content}
            </label>
            {error && <span className={errorStyles}>{error}</span>}
        </div>
    );
});

Input.displayName = "Input";
