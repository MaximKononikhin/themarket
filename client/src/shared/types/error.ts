export type IError = {
    statusCode?: number;
    message: string;
};
