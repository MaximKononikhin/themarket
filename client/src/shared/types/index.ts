export * from "./utility";
export * from "./error";
