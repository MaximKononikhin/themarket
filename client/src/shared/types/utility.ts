export type Maybe<T, S = null> = T | S;
