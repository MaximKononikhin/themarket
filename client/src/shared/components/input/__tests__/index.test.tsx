import React, { useState } from "react";

import { render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import { InputProps } from "@shared/components/input/types";

import { Input } from "../.";

interface DemoProps extends Omit<InputProps, "value" | "placeholder"> {
    placeholder?: string;
}

const PLACEHOLDER_TEXT = "Test";
const ERROR = "error";
const VALUE = "test text";
const LABEL = "test label";
const LABEL_TEST_ID = "label";

const Demo: React.FC<DemoProps> = (props) => {
    const [value, setValue] = useState("");

    const handleChange = (
        evt: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>
    ) => {
        setValue(evt.target.value);
        if (props.onChange) {
            props.onChange(evt);
        }
    };

    const placeholder = props.placeholder
        ? props.placeholder
        : PLACEHOLDER_TEXT;

    return (
        <Input
            {...props}
            value={value}
            onChange={handleChange}
            placeholder={placeholder}
        />
    );
};

describe("Input component", () => {
    it("should render input", () => {
        const handeChange = jest.fn();

        const { getByPlaceholderText, container } = render(
            <Demo onChange={handeChange} />
        );

        expect(getByPlaceholderText(PLACEHOLDER_TEXT)).toBeInTheDocument();
        expect(container).toMatchSnapshot();
    });

    it("should has proper value and called handler", async () => {
        const handeChange = jest.fn();

        const { getByPlaceholderText, container } = render(
            <Demo onChange={handeChange} />
        );

        await userEvent.type(getByPlaceholderText(PLACEHOLDER_TEXT), VALUE);

        expect(getByPlaceholderText(PLACEHOLDER_TEXT)).toHaveValue(VALUE);
        expect(handeChange).toHaveBeenCalled();
        expect(container).toMatchSnapshot();
    });

    it("should has text label", () => {
        const handeChange = jest.fn();

        const { getByText, container } = render(
            <Demo onChange={handeChange} label={LABEL} />
        );

        expect(getByText(LABEL)).toBeInTheDocument();
        expect(container).toMatchSnapshot();
    });

    it("should render element label", () => {
        const handeChange = jest.fn();

        const labelElement = <span data-testid={LABEL_TEST_ID}>label</span>;

        const { getByTestId, container } = render(
            <Demo onChange={handeChange} label={labelElement} />
        );

        expect(container).toMatchSnapshot();
        expect(getByTestId(LABEL_TEST_ID)).toBeInTheDocument();
    });

    it("should contains wide styles", () => {
        const handeChange = jest.fn();

        const { getByPlaceholderText, container } = render(
            <Demo onChange={handeChange} width="wide" />
        );

        expect(getByPlaceholderText(PLACEHOLDER_TEXT)).toHaveClass(
            "input__common_wide"
        );
        expect(container).toMatchSnapshot();
    });

    it("should render textarea", () => {
        const handeChange = jest.fn();

        const { getByPlaceholderText, container } = render(
            <Demo onChange={handeChange} design="textarea" />
        );

        expect(getByPlaceholderText(PLACEHOLDER_TEXT)).toBeInTheDocument();
        expect(container).toMatchSnapshot();
    });

    it("should render error and have proper styles", () => {
        const handeChange = jest.fn();

        const { getByText, container } = render(
            <Demo
                onChange={handeChange}
                design="textarea"
                error={ERROR}
                errorType="static"
            />
        );

        expect(getByText(ERROR)).toBeInTheDocument();
        expect(getByText(ERROR)).toHaveClass("error__static");
        expect(container).toMatchSnapshot();
    });
});
