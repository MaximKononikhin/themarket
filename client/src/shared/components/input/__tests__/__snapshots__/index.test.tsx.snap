// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input component should contains wide styles 1`] = `
<div>
  <div
    class="wrapper wrapper__wide"
  >
    <label
      class="label label__wide"
    >
      <input
        class="input input__common_wide"
        placeholder="Test"
        value=""
      />
    </label>
  </div>
</div>
`;

exports[`Input component should has proper value and called handler 1`] = `
<div>
  <div
    class="wrapper"
  >
    <label
      class="label"
    >
      <input
        class="input"
        placeholder="Test"
        value="test text"
      />
    </label>
  </div>
</div>
`;

exports[`Input component should has text label 1`] = `
<div>
  <div
    class="wrapper"
  >
    <label
      class="label"
    >
      <p
        class="textSecond"
      >
        test label
      </p>
      <input
        class="input"
        placeholder="Test"
        value=""
      />
    </label>
  </div>
</div>
`;

exports[`Input component should render element label 1`] = `
<div>
  <div
    class="wrapper"
  >
    <label
      class="label"
    >
      <span
        data-testid="label"
      >
        label
      </span>
      <input
        class="input"
        placeholder="Test"
        value=""
      />
    </label>
  </div>
</div>
`;

exports[`Input component should render error and have proper styles 1`] = `
<div>
  <div
    class="wrapper"
  >
    <label
      class="label"
    >
      <textarea
        class="textarea input__common_invalid"
        placeholder="Test"
      />
    </label>
    <span
      class="error__static"
    >
      error
    </span>
  </div>
</div>
`;

exports[`Input component should render input 1`] = `
<div>
  <div
    class="wrapper"
  >
    <label
      class="label"
    >
      <input
        class="input"
        placeholder="Test"
        value=""
      />
    </label>
  </div>
</div>
`;

exports[`Input component should render textarea 1`] = `
<div>
  <div
    class="wrapper"
  >
    <label
      class="label"
    >
      <textarea
        class="textarea"
        placeholder="Test"
      />
    </label>
  </div>
</div>
`;
